### ISSUE TITLE
Declared type of variable is ignored, leading to Never type after type narrowing

### ISSUE THREAD

**Issue by ringohoffman** [microsoft/pyright#4465]
**Describe the bug**
A variable declared as a `Union` type and with a default value that is not obviously modified... ignores the explicitly hinted type, and only takes the type of the default. This leads to a `Never` type after type narrowing.

**To Reproduce**
```python
def f() -> None:
    value: int | None = None

    def g() -> None:
        nonlocal value
        value = 1
    
    g()

    if value is None:
        return
    
    reveal_type(value)  # Pylance: Type of "value" is "Never"

```

**Expected behavior**
I would expect Pylance to use the explicitly hinted type instead only the default value's type. mypy handles this as expected.

```
mypy(note): Revealed type is "builtins.int"
```

**VS Code extension v2023.1.21**

**Comment by erictraut** [comment]
This behavior is as designed. Types of symbols are narrowed locally within an execution context. Modifications within alternate execution contexts are not considered. This is how other static type checkers (both in Python and in other languages) also work.

It's generally bad idea to rely on non-local side effects, so it's reasonable for a type checker to make this assumption. I recommend changing your code to avoid reliance on such non-local side effects.

**Comment by ringohoffman** [comment]
I am using PyTorch `Module` forward hooks to capture output layer information (dynamically detecting output size, capturing outputs for logging, ...), which is necessarily non-local, but I don't really think that is the issue here.

The issue issue is that half of my union type's hints are being discarded. I can circumvent it like this:

```python
def return_None() -> int | None:
    return None

def f() -> None:
    value: int | None = return_None()

    def g() -> None:
        nonlocal value
        value = 1
    
    g()

    if value is None:
        return
    
    reveal_type(value)  # Pylance: Type of "value" is "int"
```

but why should I have to?

**Comment by erictraut** [comment]
The type of `value` is declared as `int | None`. That means the type checker will enforce that any value assigned to it will comply with this type. The actual value of `value` at any point within `f` will depend on values assigned to it or any additional type guards that are applied. Initially, its value is narrowed to `None` based on the initial assignment. It is then narrowed to `Never` based on the `value is None` type guard. The non-local assignment performed in `g` is not considered for purposes of type analysis within `f`.

This is all working as designed.

For more details about type declarations and type inference, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md).
